llvm/tools/llvm-remarkutil/RemarkDiff.h

```diff
@@ -0,0 +1,298 @@
+/// the source file path, function name, line number and column number.
+struct DebugLocation {
+  std::string SourceFilePath;
+  std::string FunctionName;
+  unsigned SourceLine = 0;
+  unsigned SourceColumn = 0;
+  DebugLocation() = default;
+  DebugLocation(StringRef SourceFilePath, StringRef FunctionName,
+                unsigned SourceLine, unsigned SourceColumn)
+      : SourceFilePath(SourceFilePath.str()), FunctionName(FunctionName.str()),
+        SourceLine(SourceLine), SourceColumn(SourceColumn) {}
+  std::string toString() {
+    return "Ln: " + to_string(SourceLine) + " Col: " + to_string(SourceColumn);
+  }
+};
+
+/// Configure the verbosity of the diff by choosing to only show unique remarks
+/// from each version or only consider remarks if they differ in type or
+/// argument. The configurator handles user specified arguments passed by flags.
+struct DiffConfigurator {
+  bool AddRemarksFromA;
+  bool AddRemarksFromB;
+  bool ShowCommonRemarks;
+  bool ShowRemarkTypeDiff;
+  bool ShowArgTypeDiff;
+  DiffConfigurator(bool ShowArgDiffOnly, bool OnlyShowCommonRemarks,
+                   bool OnlyShowDifferentRemarks, bool ShowOnlyA,
+                   bool ShowOnlyB, bool ShowRemarkTypeDiffOnly) {
+    AddRemarksFromA = !OnlyShowCommonRemarks && (ShowOnlyA || !ShowOnlyB);
+    AddRemarksFromB = !OnlyShowCommonRemarks && (ShowOnlyB || !ShowOnlyA);
+    ShowCommonRemarks = !OnlyShowDifferentRemarks || OnlyShowCommonRemarks;
+    ShowRemarkTypeDiff = !ShowArgDiffOnly || ShowRemarkTypeDiffOnly;
+    ShowArgTypeDiff = !ShowRemarkTypeDiffOnly || ShowArgDiffOnly;
+  }
+};
+
+/// Represent a diff where the remark header information is the same but the
+/// differ in remark type or agruments.
+struct DiffAtRemark {
+  RemarkInfo BaseRemark;
+  std::optional<std::pair<Type, Type>> RemarkTypeDiff;
+  SmallVector<RemarkArgInfo, 4> OnlyA;
+  SmallVector<RemarkArgInfo, 4> OnlyB;
+  SmallVector<RemarkArgInfo, 4> InBoth;
+
+  /// Compute the diff between two remarks \p RA and \p RB which share the same
+  /// header and differ in remark type and arguments.
+  DiffAtRemark(RemarkInfo &RA, RemarkInfo &RB, DiffConfigurator &DiffConfig)
+      : BaseRemark(RA) {
+    if (DiffConfig.ShowArgTypeDiff) {
+      unsigned ArgIdx = 0;
+      // Loop through the remarks in RA and RB in order comparing both.
+      for (; ArgIdx < std::min(RA.Args.size(), RB.Args.size()); ArgIdx++) {
+        if (RA.Args[ArgIdx] == (RB.Args[ArgIdx]))
+          InBoth.push_back(RA.Args[ArgIdx]);
+        else {
+          OnlyA.push_back(RA.Args[ArgIdx]);
+          OnlyB.push_back(RB.Args[ArgIdx]);
+        }
+      }
+
+      // Add the remaining remarks if they exist to OnlyA or OnlyB.
+      SmallVector<RemarkArgInfo, 4> RemainingArgs =
+          RA.Args.size() > RB.Args.size() ? RA.Args : RB.Args;
+      bool IsARemaining = RA.Args.size() > RB.Args.size() ? true : false;
+      for (; ArgIdx < RemainingArgs.size(); ArgIdx++)
+        if (IsARemaining)
```

Comment:
Is it possible to do this?

`auto &RemainingRemarks = RA.Args.size() ? RB.Args.size() ? RA : RB;`

Reply:
The code I have copies the rest of the remaining remarks into OnlyA or OnlyB. I guess I can move the check of `IsARemaining` outside the loop